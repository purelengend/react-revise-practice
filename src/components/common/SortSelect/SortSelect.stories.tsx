import type { Meta, StoryObj } from "@storybook/react";
import { Center } from "@chakra-ui/react";

// Components
import SortSelect from ".";

// Constants
import { SORT_BY_OPTION_LIST } from "@/constants";

const meta: Meta<typeof SortSelect> = {
  title: "Components/SortSelect",
  component: SortSelect,
  decorators: [
    (Story) => (
      <Center style={{ width: "100vh" }}>
        <Story />
      </Center>
    ),
  ],
  argTypes: {
    sortList: {
      description: "The list of selection options for sorting",
    },
  },
  args: {
    sortList: SORT_BY_OPTION_LIST,
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof SortSelect>;

export const Primary: Story = {};
