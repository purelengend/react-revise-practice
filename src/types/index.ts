export * from "./student";
export * from "./role";
export * from "./sidebar";
export * from "./sort";
