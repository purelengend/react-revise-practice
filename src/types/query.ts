export type StudentQuery = {
  page: number;
  limit: number;
  sortBy: string;
  order: string;
  name: string;
};
