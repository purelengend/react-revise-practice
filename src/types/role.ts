import { ROLES } from "@/constants";

export type Role = (typeof ROLES)[keyof typeof ROLES];
