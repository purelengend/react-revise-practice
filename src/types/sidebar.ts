type SideBarItem = {
  title: string;
  icon: React.ReactNode;
  destination: string;
};
export type SideBarItemList = Array<SideBarItem>;
