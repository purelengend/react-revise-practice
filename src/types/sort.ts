export type SortProps = {
  title: string;
  value: string;
};
