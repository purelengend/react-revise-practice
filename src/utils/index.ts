export * from "./test-utils";
export * from "./config";
export * from "./customToast";
