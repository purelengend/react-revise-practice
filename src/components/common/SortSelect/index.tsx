import { memo, useCallback, useEffect, useMemo, useState } from "react";
import { Button, Select } from "@chakra-ui/react";

// Components
import { SortArrowDownIcon, SortArrowUpIcon } from "../Icons";

// Types
import { SortProps } from "@/types";

// Constants
import { QUERY_PARAMS } from "@/constants";

// Hooks
import { useQueryParams } from "@/hooks";

const SortSelect = memo((props: { sortList: Array<SortProps> }) => {
  const [isAscending, setIsAscending] = useState(false);

  const { searchParams, setSearchParams, sortBy } = useQueryParams();

  // Set default sort parameter
  useEffect(() => {
    if (!sortBy) {
      searchParams.set(
        QUERY_PARAMS.SORTBY,
        props.sortList[props.sortList.length - 1].value,
      );
      setSearchParams(searchParams);
    }
  }, [props.sortList, searchParams, setSearchParams, sortBy]);

  useEffect(() => {
    isAscending
      ? searchParams.set(QUERY_PARAMS.ORDER, "asc")
      : searchParams.set(QUERY_PARAMS.ORDER, "desc");

    setSearchParams(searchParams);
  }, [isAscending, searchParams, setSearchParams, sortBy]);

  const SortByStateIcon = useMemo(
    () =>
      isAscending ? (
        <SortArrowUpIcon data-testid="asc" />
      ) : (
        <SortArrowDownIcon data-testid="desc" />
      ),
    [isAscending],
  );

  const orderCallback = useCallback(() => setIsAscending((prev) => !prev), []);

  const sortCallback = useCallback(
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      if (e.target.value) {
        searchParams.set(QUERY_PARAMS.SORTBY, e.target.value);
      }
      setSearchParams(searchParams);
    },
    [searchParams, setSearchParams],
  );

  const defaultSortValue = props.sortList[props.sortList.length - 1].value;

  return (
    <>
      <Button
        type="button"
        bg="transparent"
        _hover={{
          bg: "blackAlpha.100",
        }}
        onClick={orderCallback}
        data-testid="sort-button"
        aria-label="sort-button"
      >
        {SortByStateIcon}
      </Button>

      <Select
        w={{
          base: "full",
          sm: 45,
        }}
        h={11}
        color="gray.200"
        fontSize="sm"
        onChange={sortCallback}
        aria-label="sort-select"
        defaultValue={defaultSortValue}
      >
        {props.sortList.map((sort) => (
          <option key={sort.title} value={sort.value}>
            {sort.title}
          </option>
        ))}
      </Select>
    </>
  );
});

export default SortSelect;
