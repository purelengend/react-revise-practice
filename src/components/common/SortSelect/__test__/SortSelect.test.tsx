import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react";

// Component
import SortSelect from "..";

// Types
import { SortProps } from "@/types";

// Utils
import { customRender } from "@/utils";

describe("SortSelect test cases", () => {
  const mockSortList: Array<SortProps> = [
    {
      title: "mock title",
      value: "mock value",
    },
  ];

  const setup = () => {
    userEvent.setup();

    return customRender(<SortSelect sortList={mockSortList} />);
  };

  it("should render correctly", async () => {
    const { container } = await act(() => setup());

    expect(container).toMatchSnapshot();
  });

  it("should render with default sort icon", async () => {
    await act(() => setup());

    const defaultSortIcon = screen.getByTestId("default");

    expect(defaultSortIcon).toBeInTheDocument();
  });

  it("should render ascending sort icon when selecting by default", async () => {
    await act(() => setup());

    const selectCombobox = screen.getByRole("combobox");

    const selectOption = screen.getByRole("option", { name: "mock title" });

    await userEvent.selectOptions(selectCombobox, selectOption);

    const ascendingSortIcon = screen.getByTestId("asc");

    expect(ascendingSortIcon).toBeInTheDocument();
  });

  it("should toggle state of sort icon", async () => {
    await act(() => setup());

    const selectCombobox = screen.getByRole("combobox");

    const selectOption = screen.getByRole("option", { name: "mock title" });

    await userEvent.selectOptions(selectCombobox, selectOption);

    const ascendingSortIcon = screen.getByTestId("asc");

    expect(ascendingSortIcon).toBeInTheDocument();

    await userEvent.click(ascendingSortIcon);

    const descendingSortIcon = screen.getByTestId("desc");

    expect(descendingSortIcon).toBeInTheDocument();
  });
});
