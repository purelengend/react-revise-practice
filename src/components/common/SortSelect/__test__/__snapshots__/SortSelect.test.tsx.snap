// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortSelect test cases should render correctly 1`] = `
<div>
  <button
    aria-label="sort-button"
    class="chakra-button css-wbul9j"
    data-testid="sort-button"
    type="button"
  >
    <svg
      class="chakra-icon css-1ad70y2"
      data-testid="desc"
      focusable="false"
      viewBox="0 0 17 22"
    >
      <path
        d="M15.6235 9L1.37652 9C0.986507 9 0.659356 8.91111 0.395065 8.73333C0.130774 8.55556 -0.000912914 8.34815 4.76302e-06 8.11111C4.76302e-06 8.03704 0.016982 7.95911 0.0509361 7.87733C0.0848901 7.79556 0.136739 7.71793 0.206482 7.64445L7.32995 0.400008C7.4676 0.266675 7.63966 0.166526 7.84614 0.0995626C8.05262 0.0325994 8.27057 -0.000584602 8.49998 7.62939e-06C8.7294 0.000600815 8.94735 0.0337849 9.15383 0.0995626C9.36031 0.16534 9.53237 0.26549 9.67002 0.400008L16.7935 7.64445C16.8623 7.71852 16.9142 7.79615 16.949 7.87733C16.9839 7.95852 17.0009 8.03645 17 8.11111C17 8.34815 16.8678 8.55556 16.6035 8.73333C16.3392 8.91111 16.0125 9 15.6235 9ZM13.1113 7.22222L8.49998 2.55556L3.88866 7.22222L13.1113 7.22222Z"
        fill="currentColor"
      />
      <path
        d="M7.4622 21.5691L0.296512 14.6277C0.197674 14.5319 0.123876 14.4285 0.0751164 14.3174C0.0263568 14.2064 0.00131783 14.0864 0 13.9574C0 13.7021 0.0909302 13.4787 0.27279 13.2872C0.454651 13.0957 0.693178 13 0.988371 13H16.0116C16.3081 13 16.5473 13.0957 16.7292 13.2872C16.911 13.4787 17.0013 13.7021 17 13.9574C17 14.0213 16.9011 14.2447 16.7035 14.6277L9.53778 21.5691C9.37305 21.7287 9.20833 21.8404 9.0436 21.9043C8.87887 21.9681 8.69767 22 8.49999 22C8.30232 22 8.12112 21.9681 7.95639 21.9043C7.79166 21.8404 7.62693 21.7287 7.4622 21.5691Z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    class="chakra-select__wrapper css-1vquela"
  >
    <select
      aria-label="sort-select"
      class="chakra-select css-mdekdx"
    >
      <option
        selected=""
        value="mock value"
      >
        mock title
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-1td5bvu"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
